File: linear_regression.R
# load the necessary libraries
library(readr)
library(dplyr)
library(caret)
library(car)
library(ggplot2)
library(MASS)

# read in your dataset
df <- read_csv("C:/tera reynolds/techscore.csv")
head(df)
# set the independent variables (X) and dependent variable (y)
X <- subset(df, select = c(otherPHR, PP, thirdParty, miPHR, selfTrackingApp, homeMedDevice, wearable))
y <- df$techScore

# split the dataset into train and test sets with a 70-30 ratio
set.seed(42)
train_index <- createDataPartition(y, p = 0.8, list = FALSE)
X_train <- X[train_index, ]
X_test <- X[-train_index, ]
y_train <- y[train_index]
y_test <- y[-train_index]

# fit the multiple linear regression model on the training set
model <- lm(y_train ~ ., data = cbind(X_train, y_train))

# predict the response variable on the test set
y_pred <- predict(model, newdata = X_test)

# calculate the mean squared error and mean absolute error
mse <- mean((y_test - y_pred)^2)
mae <- mean(abs(y_test - y_pred))

print(paste("Mean Squared Error:", mse))
print(paste("Mean Absolute Error:", mae))

# print the model summary
summary(model)

# create a data frame of predicted and actual values
residuals_df <- data.frame(y_pred = y_pred, y_actual = y_test, residual = y_test - y_pred)

# plot residuals vs predicted values
ggplot(residuals_df, aes(x = y_pred, y = residual)) +
  geom_point() +
  geom_hline(yintercept = 0, linetype = "dashed") +
  xlab("Predicted Values") +
  ylab("Residuals") +
  ggtitle("Residuals vs Predicted Values")

# create a data frame with predicted and actual values
df_results <- data.frame(y_actual = y_test, y_predicted = y_pred)

# plot predicted vs. actual values
ggplot(df_results, aes(x = y_predicted, y = y_actual)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, color = "red") +
  labs(x = "Predicted values", y = "Actual values", title = "Predicted vs. Actual values")

File: linear_regression.py
# @Time    : 4/2/2023 2:44 PM
# @Author  : Pronob Barman
# @FileName: linear_regression.py
# @Software: PyCharm

'''
This script is used to perform linear regression on the techScore data
'''

# import libraries
import pandas as pd
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score


# read csv file
data = pd.read_csv('linear_regression.csv')

# print(data.head())

# print(data.columns)

# drop rows with missing values
data.dropna(inplace=True)

# create dummy variables for the 'race' column
# race_dummies = pd.get_dummies(data['race'], prefix='race')
# concatenate the dummy variables with the original data
# data = pd.concat([data, race_dummies], axis=1)

# drop the original 'race' column
data.drop('race', axis=1, inplace=True)

# response variable
y = data['techScore']
# predictor variables
x = data.iloc[:, 1:]
print(x.head())
# split the data into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

# fit the model
regressor = LinearRegression()
model = regressor.fit(x_train, y_train)

# print the coefficients
print("Model Coefficient:", model.coef_)
print("Model Intercept: ", model.intercept_)
# print the score or r-squared value
print("R-squared value: ", model.score(x_test, y_test))
# print the predictions for the test set
y_pred = model.predict(x_test)
print("y predicted values: ", y_pred)

# generate and print model summary using statsmodels
X_train = sm.add_constant(x_train) # add a constant term for intercept
model_sm = sm.OLS(y_train, x_train).fit()
print(model_sm.summary())

# add constant term to test data
X_test = sm.add_constant(x_test)
# predict response variable on test data using fitted model
y_pred = model_sm.predict(x_test)
r2 = r2_score(y_test, y_pred)
print("R-squared value: ", r2)
# print intercept and coefficients
print('Intercept:', model_sm.params[0])
print('Coefficients:')
print(model_sm.params[1:])

# generate and print model summary
print(model_sm.summary())


# plot the residuals vs predicted values
plt.scatter(y_pred, y_pred - y_test)
plt.xlabel('Predicted Values')
plt.ylabel('Residuals')
plt.title('Residuals vs Predicted Values')
plt.show()

# create a file object
with open('linear_regression_results.txt', 'w') as f:
    # write the model coefficients to the file
    f.write('Model Coefficients: {}\n'.format(model.coef_))

    # write the model intercept to the file
    f.write('Model Intercept: {}\n'.format(model.intercept_))

    # write the r-squared value to the file
    # f.write('R-squared value: {}\n'.format(model.score(x_test, y_test)))

    # write the predictions for the test set to the file
    # f.write('Predicted Values: {}\n'.format(y_pred))
    # write the intercept to the file
    f.write('Intercept:{}\n'.format(model_sm.params[0]))
    # write the model summary to the file
    f.write('{}\n'.format(model_sm.summary()))
